### PR TITLE
fix(core/router): avoid compare same table of router's destinations

### LINKED ISSUE #10446: Error when configuring routes with more than three destination rules
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

2.4、2.5

### Current Behavior

When there are more than 3 destination rules in the routing configuration, a configuration loading error will occur. The error is: 
```
invalid order function for sorting.
```

Route config is ： 
```
{"next":null,"data":[{"name":"rrr1","path_handling":"v0","id":"460a58d8-5701-49cf-8cc7-a714195b812e","regex_priority":0,"destinations":[{"ip":"0.0.0.0/0","port":7000},{"ip":"0.0.0.0/0","port":7001},{"ip":"0.0.0.0/0","port":7002},{"ip":"0.0.0.0/0","port":7003},{"ip":"0.0.0.0/0","port":7004}],"paths":null,"created_at":1678187654,"updated_at":1678187654,"headers":null,"methods":null,"sources":null,"tags":["9b73b55e-238c-4d1d-b10e-cc06417071fa"],"protocols":["tcp"],"strip_path":true,"request_buffering":true,"response_buffering":true,"snis":null,"hosts":null,"service":{"id":"9b73b55e-238c-4d1d-b10e-cc06417071fa"},"https_redirect_status_code":426,"preserve_host":false}]}
```

The error should be that the destination rule comparison function does not return a stable result.
```
local function sort_destinations(r1, _)
  for _, destination in ipairs(r1.destinations) do
    if destination.ip and destination.port then
      return true
    end
  end
end
```

### Expected Behavior

Configure more than 3 destination rules without error.

### Steps To Reproduce

Create a route for rules with more than 3 destination rules. 

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by sabertobihwy**
### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix #10446

KAG-918

### CODE REVIEW

**Comment by chronolaw** [comment]
We should add a change log entry in `fixes` section.

File: CHANGELOG.md [hunk 1]
```diff
@@ -85,6 +85,9 @@
 
 ### Fixes
 
+- Avoid compare same table of router's destinations
+  [#10470](https://github.com/Kong/kong/pull/10470)
```
Comment (chronolaw) [inline CHANGELOG.md:89]: should in `Core` section, sorry for my previous unclear comment.

Comment (chronolaw) [reply CHANGELOG.md:89]: The description should be `Fixed an issue ...`

**Comment by chronolaw** [comment]
> @sabertobihwy我试过你的修改，但是当路由不止一条的时候，还是会报错。重现如下：
> 
> ```
> 
> curl -sS --fail  \
>      -X POST http://localhost:8001/services/test_svc_tcp/routes \
>      -H 'Content-Type: application/json' \
>      -d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":5353},{"ip":"0.0.0.0/0","port":5354},{"ip":"0.0.0.0/0","port":5355},{"ip":"0.0.0.0/0","port":5356}]}'
> 
> 
> curl -sS --fail  \
> -X POST http://localhost:8001/services/test_svc_tcp/routes \
> -H 'Content-Type: application/json' \
> -d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":7001}]}'
> ```

It seems that we have a incorrect test case, @sabertobihwy could you check it again?

**Comment by chronolaw** [comment]
I think function `sort_sources()` may have the same problem, could we check it again?

**Comment by zhangzerui20** [comment]
@chronolaw It seems that sort_destinations doesn't really compare r1 and r2.

How do we determine the priority of routes by destinations? I think the possible strategy is that if r1 has a rule that includes both ip and port, but r2 does not, then r1 has a higher priority. In other cases, compare the creation time of r1 and r2.

Is this reasonable?

**Comment by sabertobihwy** [comment]
Hi, let’s discuss the sorting rule for `route.sources` and `route.destinations`. Here are some guidelines:

1. If there is only one route, it doesn’t need to be sorted.

2. If there are two routes, we need to determine how to sort them, routes like:

	- r1:
	```
	    destinations = { {
	        ip = "0.0.0.0",
	        port = 19004
	      }, {
	        ip = "0.0.0.0",
	        port = 19005
	      }, {
	        ip = "0.0.0.0",
	        port = 19006
	      }, {
	        ip = "0.0.0.0",
	        port = 19007
	      },
	      [0] = 4
	    },
	    headers = {
	      [0] = 0
	    },
	    hosts = {
	      [0] = 0
	    },
	```
		
	- r2:
	
	```
	    destinations = { {
	        ip = "0.0.0.0",
	        port = 19008
	      },
	      [0] = 1
	    },
	    headers = {
	      [0] = 0
	    },
	    hosts = {
	      [0] = 0
	    },
	```

The original sorting rule was to prioritize routes with the full IP and port, but this fails in cases like r1 and r2. Thus, we need to reconsider the sort rules, such as routes' `updated_at`.

**Comment by chronolaw** [comment]
@bungle , could you give us some comments?

File: kong/router/traditional.lua [hunk 2]
```diff
@@ -719,8 +720,17 @@ local function sort_sources(r1, _)
 end
 
 
-local function sort_destinations(r1, _)
+local function sort_destinations(r1, r2)
   local destinations = r1.destinations
+
+  local r1_dst_addr = tostring(destinations)
+  local r2_dst_addr = tostring(r2.destinations)
```
[the remark this answers is not in the document]
Comment (bungle) [reply kong/router/traditional.lua:727]: @sabertobihwy ^ I think this should be fine, can you make the change? Otherwise I think it is alright.

**Comment by bungle** [comment]
> @chronolaw It seems that sort_destinations doesn't really compare r1 and r2.
> 
> How do we determine the priority of routes by destinations? I think the possible strategy is that if r1 has a rule that includes both ip and port, but r2 does not, then r1 has a higher priority. In other cases, compare the creation time of r1 and r2.
> 
> Is this reasonable?

I am fine with adding:
```
  if r1.route.created_at ~= nil and r2.route.created_at ~= nil then
    return r1.route.created_at < r2.route.created_at
  end
  ```

to that sort.

I guess the original goal was to try the `ip, port` combinations first. But if I look it correctly it is not that. It looks like if there is at least one source/destination with both then those are sorted first. Is it more correct to try all those ip port combinations first before trying just with ip and port, and even there we can decide if ip should be matched before port.

**Comment by zhangzerui20** [comment]
> > @chronolaw It seems that sort_destinations doesn't really compare r1 and r2.
> > How do we determine the priority of routes by destinations? I think the possible strategy is that if r1 has a rule that includes both ip and port, but r2 does not, then r1 has a higher priority. In other cases, compare the creation time of r1 and r2.
> > Is this reasonable?
> 
> I am fine with adding:
> 
> ```
>   if r1.route.created_at ~= nil and r2.route.created_at ~= nil then
>     return r1.route.created_at < r2.route.created_at
>   end
> ```
> 
> to that sort.
> 
> I guess the original goal was to try the `ip, port` combinations first. But if I look it correctly it is not that. It looks like if there is at least one source/destination with both then those are sorted first. Is it more correct to try all those ip port combinations first before trying just with ip and port (across all the routes), and even there we can decide if ip should be matched before port.
> 
> The original sorting thing dates 5 years back: [acec799#diff-1958886e00e07e58fb444abe5b024fa98ba0c67c902d4c017fb41d97928de428R880-R900](https://github.com/Kong/kong/commit/acec7995c67dfe1cad22360310fcf7525ee56a69#diff-1958886e00e07e58fb444abe5b024fa98ba0c67c902d4c017fb41d97928de428R880-R900)
> 
> Quite hard to understand what it tries to do. It probably has something to do with these tests: [acec799#diff-d728734c307eca97c7e5d00248bf644ab30c1aae30102361971396715e972b1dR2044-R2062](https://github.com/Kong/kong/commit/acec7995c67dfe1cad22360310fcf7525ee56a69#diff-d728734c307eca97c7e5d00248bf644ab30c1aae30102361971396715e972b1dR2044-R2062)
> 
> Let's try to write that open (sorting sources):
> 
> ```lua
> if not isempty(categories) then
>   for _, category in pairs(categories) do
>     local source = category.routes_by_sources
>     if not isempty(source) then
>       for _, routes in pairs(source) do -- TODO: why does it use `pairs` here?
>         sort(routes, function sort_sources(r1, _)
>           local sources = r1.sources
>           for i = 1, sources[0] do
>             if sources[i].ip and sources[i].port then
>               return true
>             end
>           end
>         end)
>       end
>     end
>   end
> end
> ```
> 
> So what does it try to do? How does that affect what it is doing later?
> 
> 1. there has to be categories
> 2. there has to be `routes_by_sources` within (each) category
> 3. the `routes_by_sources` contains elements that are `routes` (?)
> 4. the `routes` are then sorted
> 
> The 3 and 4 I cannot wrap my head around without looking what is inside those data structures. For me it feels like it tries to sort these ip-port records inside a single route (not trying to sort between different routes).

@bungle 
sources and destinations should be the same, I read the code related to destinations. The structure of routes_by_destinations is like this.

```
routes_by_destinations = {
    {
        -- router_1 : "destinations":[{"ip":"0.0.0.0/0", "port":7000}]
        -- router_2 : "destinations":[{"ip":"0.0.0.0/0", "port":7001}]
        -- router_3 : "destinations":[{"ip":"0.0.0.0/0", "port":7002}]

        "0.0.0.0/0" = [router_1, router_2, router_3],
        "7000" = [route_1], 
        "7001" = [route_2],
        "7002" = [route_3]
    }
}
```

The reduce() function in the find_match() function will use the router list returned by routes_by_ destinations, to match request.

So it looks like it should match the router, not the ip-port.

**Comment by sabertobihwy** [comment]
> For me it feels like it tries to sort these ip-port records inside a single route (not trying to sort between different routes).

> So it looks like it should sort the router, not the ip-port.

Now, the code actually sorts the routes, not the ip-port in the individual routes.

I have a suggestion: 

1. For route sorting, choose to sort by `created_at` (does Kong have a unified rule for sorting by `created_at`, in ascending or descending order?).
2. For the `src` and `dst` of a single route, choose to sort by IP-port (same as before, IP+port > only IP > only port).

Focus on this PR: I want to handle it as sorting the ip-port of a single route.

cc @bungle @zhangzerui20 WDYT

**Comment by zhangzerui20** [comment]
> Focus on this PR: I want to handle it as sorting the ip-port of a single route.

I think that's fine. The sorting of routes should all be placed in sort_routes(). For this problem, the sorting of src and dst is here.

**Comment by bungle** [comment]
Ok, cool:

So it pairs over this:
```
 {
        "0.0.0.0/0" = [route_1, route_2, route_3],
        "7000" = [route_1, route_1], 
        "7001" = [route_2],
        "7002" = [route_3]
    }
```

where route_1 looks like this:

```
{
  destinations = {
    {
      ip = "0.0.0.0/0",
      port = 7000,
    },
    {
      port = 7000,
    },
  }
}
```

which means that the same `route_1` is mentioned twice in `"7000"`. I think it would be a way better to not put the `route_1` twice to `"7000"`.

Again if I understood correctly, so I would fix it so that it is not appended here twice to "7000":
https://github.com/Kong/kong/blob/master/kong/router/traditional.lua#L766-L783

**Comment by bungle** [comment]
@sabertobihwy / @zhangzerui20,

Now that I understood it better, I created alternate implementation:
https://github.com/Kong/kong/pull/10514

Let me know your thoughts on it.

**Comment by zhangzerui20** [comment]
@bungle 

I think the cause of the problem is not that the same elements appear in the sort() array (this is another problem). The problem is that lua's table.sort method requires passing in an order function, `This order function receives two arguments and must return true if the first argument should come first in the sorted array` (refer to [lua doc](https://www.lua.org/pil/19.3.html).)

When the destinations of all routes contain both ip and port, our sort_destinations() cannot sort two routes. When sort_destinations() is passed in `route_1` and `route_2`, sort_destinations() returns `true`, and when sort_destinations() is passed in `route_2` and `route_1` also returns `true`. This causes confusion in lua's table.sort(), lua's table.sort() returns "invalid order function for sorting".

Therefore, when the length of an array of routes_by_destinations exceeds `3`, even if there are no repeated routes, an error will be reported `invalid order function for sorting`. For example, routes_by_destinations below, when sorting the array corresponding to 0.0.0.0/0, an error will be reported.
```
routes_by_destinations = {
        "0.0.0.0/0" = [route_1, route_2, route_3, route_4],
        ... ...
}
```

Why are there more than `3` items? I am not very familiar with lua. I looked at the source code of table.sort(), and I can see that sort does special processing for 3 elements.(refer to [lua sort source code](https://www.lua.org/source/5.2/ltablib.c.html))

image

So, I think we need to fix sort_destinations() , or not sort routes but sort src and dst in a single route.

**Comment by bungle** [comment]
> I think the cause of the problem is not that the same elements appear in the sort() array (this is another problem). The problem is that lua's table.sort() function requires passing in an order function, `This order function receives two arguments and must return true if the first argument should come first in the sorted array` (refer to [lua doc](https://www.lua.org/pil/19.3.html).)
> 
> When the destinations of all routes contain both ip and port, our sort_destinations() cannot sort two routes. When sort_destinations() is passed in `route_1` and `route_2`, sort_destinations() returns `true`, and when sort_destinations() is passed in `route_2` and `route_1` also returns `true`. This causes confusion in lua's table.sort(), lua's table.sort() returns `invalid order function for sorting`.
> 
> Therefore, when the length of an array of routes_by_destinations exceeds `3`, even if there are no same routes, an error will be reported `invalid order function for sorting

I think the confusion is caused only when the same table is there more than once. Otherwise it just depends on order sort processes them. Both of your examples are valid sorts.

**Comment by zhangzerui20** [comment]
> We need to double check it though. Fo you have test case in your mind that could prove your point?

Create 4 routes to reproduce.

```
curl -i -s -X POST http://localhost:8001/services \
  --data name=test_svc_tcp \
  --data host='mockbin.org'\
  --data port='8080' \
  --data protocol='tcp'

curl -sS --fail  \
-X POST http://localhost:8001/services/test_svc_tcp/routes \
-H 'Content-Type: application/json' \
-d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":7000}]}'

curl -sS --fail  \
-X POST http://localhost:8001/services/test_svc_tcp/routes \
-H 'Content-Type: application/json' \
-d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":7001}]}'

curl -sS --fail  \
-X POST http://localhost:8001/services/test_svc_tcp/routes \
-H 'Content-Type: application/json' \
-d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":7002}]}'

curl -sS --fail  \
-X POST http://localhost:8001/services/test_svc_tcp/routes \
-H 'Content-Type: application/json' \
-d '{"protocols":["tcp"],"destinations":[{"ip":"0.0.0.0/0","port":7003}]}'

```

**Comment by bungle** [comment]
> Create 4 routes to reproduce.

You are right. I could reproduce this error now with PR #10514. I needed to do your curl and then make request to stream port to see it.

**Comment by bungle** [comment]
Hi @zhangzerui20 / @sabertobihwy,

I made some changes to #10514, and it doesn't seem to error anymore. I am still a bit lost on what this sorting or `route_by_destionations/-sources` happens to do? If a single route (entity) has rule that contains `ip` and `port` it is sorted before those that don't, but at the same time same route can have rules that have `ip` AND `port`, or just `ip` or `port`.

**Comment by zhangzerui20** [comment]
So should it be that routes with only `ip` and `port` have higher priority?

**Pull request closed without merging**